tools/trait_validity/check.py

```diff
@@ -0,0 +1,107 @@
+how_to_fix_message = "Please ensure that all traits in the code/__DEFINES/trait_declarations.dm file are added in the code/_globalvars/traits.dm file."
+
+def green(text):
+	return "\033[32m" + str(text) + "\033[0m"
+
+def red(text):
+	return "\033[31m" + str(text) + "\033[0m"
+
+def blue(text):
+	return "\033[34m" + str(text) + "\033[0m"
+
+# simple way to check if we're running on github actions, or on a local machine
+on_github = os.getenv("GITHUB_ACTIONS") == "true"
+
+defines_file = "code/__DEFINES/traits/declarations.dm"
+globalvars_file = "code/_globalvars/traits.dm"
+
+def post_error(define_name):
+	if on_github:
+		print(f"::error file={defines_file},title=Define Sanity::{define_name} is defined in {defines_file} but not added to {globalvars_file}!")
+	else:
+		print(red(f"- Failure: {define_name} is defined in {defines_file} but not added to {globalvars_file}!"))
+
+number_of_defines = 0
+
+if not on_github:
+	print(blue(f"Running define sanity check outside of Github Actions.\nFor assistance, a '{output_file_name}' file will be generated at the root of your directory if any errors are detected."))
+
+if not os.path.isfile(defines_file):
+	print(red(f"Could not find the defines file '{defines_file}'!"))
+	sys.exit(1)
+
+if not os.path.isfile(globalvars_file):
+	print(red(f"Could not find the globalvars file '{globalvars_file}'!"))
+	sys.exit(1)
+
+defines_to_search_for = []
+missing_defines = []
+scannable_lines = []
+
+with open(defines_file, 'r') as file:
+	reading = False
+
+	for line in file:
+		line = line.strip()
+
+		if line == "// BEGIN TRAIT DEFINES":
+			reading = True
+			continue
+		elif line == "// END TRAIT DEFINES":
+			break
+		elif not reading:
+			continue
+
+		scannable_lines.append(line)
+
+for potential_define in scannable_lines:
+	match = define_regex.match(potential_define)
+	if not match:
+		continue
+
+	number_of_defines += 1
+	defines_to_search_for.append(match.group(2))
+
+if number_of_defines == 0:
+	print(red("No defines found! This is likely an error."))
+	sys.exit(1)
+
+if number_of_defines <= 450:
+	print(red(f"Only found {number_of_defines} defines! Something has likely gone wrong as the number of local defines should not be this low."))
+	sys.exit(1)
+
+with open(globalvars_file, "r") as file:
+	globalvars_file_contents = file.read()
+	for define_name in defines_to_search_for:
+		searchable_string = "\"" + define_name + "\" = " + define_name
+		if not re.search(searchable_string, globalvars_file_contents):
+			missing_defines.append(define_name)
+
+if len(missing_defines):
+	string_list = []
+	for missing_define in missing_defines:
+		if not on_github:
+			post_error(missing_define)
+			string_list.append(f"{missing_define} is defined in {defines_file} but not added to {globalvars_file}!")
+		else:
+			post_error(missing_define)
+
+	if len(string_list):
+		with open(output_file_name, "w") as output_file:
+			output_file.write("\n".join(string_list))
+			output_file.write("\n\n" + how_to_fix_message)
```

Comment:
I'm not sure we have any need to put this in an output file, just dump it to stderr.

Reply:
this was for me so i could quickly view the txt file and fix the codebase without scrolling for aeons through the terminal, also probably useful to other people who port this tool and need to do the mass change.

Reply:
if you want it in a file, then `>` into a file. if you want to read through large outputs easily, pipe into `less` or `bat`. but it's not right for this to output into a file itself

Reply:
i could not find any information on using `>` in python, nor `less` neither `bat`. could you link to an example for any of those three solutions? it seems the only way to write to a file is creating a file in python right now unless my research has been faulty someone (even `logger` module requires you do something similar to open up a file).

Reply:
oh, do you want it to be either/or on the terminal versus the file? that's something that I don't really like because it's really simple to jump to the file reference when you run it in the vscode integrated terminal and see what the issue is while also having the text file handy to track ongoing progress.

i've already done all the work i needed to so i'll just remove this part of the code.